### ISSUE TITLE
10 Examples of Linear Algebra in Machine Learning

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#502]
10 Examples of Linear Algebra in Machine Learning<br>
<br>
https://ift.tt/2If4OwQ<br>
<br>
Jason Brownlee<br>
<br>
<div>
<section class="entry">
<p>Linear algebra is a sub-field of mathematics concerned with vectors, matrices, and linear transforms.</p>
<p>It is a key foundation to the field of machine learning, from notations used to describe the operation of algorithms to the implementation of algorithms in code.</p>
<p>Although linear algebra is integral to the field of machine learning, the tight relationship is often left unexplained or explained using abstract concepts such as vector spaces or specific matrix operations.</p>
<p>In this post, you will discover 10 common examples of machine learning that you may be familiar with that use, require and are really best understood using linear algebra.</p>
<p>After reading this post, you will know:</p>
<ul>
<li>The use of linear algebra structures when working with data, such as tabular datasets and images.</li>
<li>Linear algebra concepts when working with data preparation, such as one hot encoding and dimensionality reduction.</li>
<li>The ingrained use of linear algebra notation and methods in sub-fields such as deep learning, natural language processing, and recommender systems.</li>
</ul>
<p>Let’s get started.</p>
<div id="attachment_4944" class="wp-caption aligncenter">
<img class="size-full wp-image-4944" src="https://3qeqpr26caki16dnhd19sv6by6v-wpengine.netdna-ssl.com/wp-content/uploads/2018/03/10-Examples-of-Linear-Algebra-in-Machine-Learning.jpg" alt="10 Examples of Linear Algebra in Machine Learning"><p class="wp-caption-text">10 Examples of Linear Algebra in Machine Learning<br>Photo by <a href="https://www.flickr.com/photos/jsbarbosa/33573815251/">j. Barbosa</a>, some rights reserved.</p>
</div>
<h2>Overview</h2>
<p>In this post, we will review 10 obvious and concrete examples of linear algebra in machine learning.</p>
<p>I tried to pick examples that you may be familiar with or have even worked with before. They are:</p>
<ol>
<li>Dataset and Data Files</li>
<li>Images and Photographs</li>
<li>One-Hot Encoding</li>
<li>Linear Regression</li>
<li>Regularization</li>
<li>Principal Component Analysis</li>
<li>Singular-Value Decomposition</li>
<li>Latent Semantic Analysis</li>
<li>Recommender Systems</li>
<li>Deep Learning</li>
</ol>
<p>Do you have your own favorite example of linear algebra in machine learning?<br>
Let me know in the comments below.</p>
<p></p>
<p class="woo-sc-hr"></p>
<center>
<h3>Need help with Linear Algebra for Machine Learning?</h3>
<p>Take my free 7-day email crash course now (with sample code).</p>
<p>Click to sign-up and also get a free PDF Ebook version of the course.</p>
<p><a href="https://machinelearningmastery.lpages.co/leadbox/143e9e9f3f72a2%3A164f8be4f346dc/5667039158992896/" target="_blank">Download Your FREE Mini-Course</a></p>
</center>
<p></p>
<p class="woo-sc-hr"></p>
<h2>1. Dataset and Data Files</h2>
<p>In machine learning, you fit a model on a dataset.</p>
<p>This is the table-like set of numbers where each row represents an observation and each column represents a feature of the observation.</p>
<p>For example, below is a snippet of the <a href="http://archive.ics.uci.edu/ml/datasets/Iris">Iris flowers dataset</a>:</p>

		<div id="crayon-5aa76af7156d8562847668" class="crayon-syntax crayon-theme-classic crayon-font-monaco crayon-os-pc print-yes notranslate" data-settings=" minimize scroll-mouseover">
		
			
			<p class="crayon-info"></p>
			<p class="crayon-plain-wrap"></p>
			<div class="crayon-main">
				<table class="crayon-table">
					<tr class="crayon-row">
				<td class="crayon-nums " data-settings="show">
					
				</td>
						<td class="crayon-code"><div class="crayon-pre">
<p class="crayon-line" id="crayon-5aa76af7156d8562847668-1">5.1,3.5,1.4,0.2,Iris-setosa</p>
<p class="crayon-line crayon-striped-line" id="crayon-5aa76af7156d8562847668-2">4.9,3.0,1.4,0.2,Iris-setosa</p>
<p class="crayon-line" id="crayon-5aa76af7156d8562847668-3">4.7,3.2,1.3,0.2,Iris-setosa</p>
<p class="crayon-line crayon-striped-line" id="crayon-5aa76af7156d8562847668-4">4.6,3.1,1.5,0.2,Iris-setosa</p>
<p class="crayon-line" id="crayon-5aa76af7156d8562847668-5">5.0,3.6,1.4,0.2,Iris-setosa</p>
</div></td>
					</tr>
				</table>
			</div>
		</div>

<p>This data is in fact a matrix: a key data structure in linear algebra.</p>
<p>Further, when you split the data into inputs and outputs to fit a supervised machine learning model, such as the measurements and the flower species, you have a matrix (X) and a vector (y). The vector is another key data structure in linear algebra.</p>
<p>Each row has the same length, i.e. the same number of columns, therefore we can say that the data is vectorized where rows can be provided to a model one at a time or in a batch and the model can be pre-configured to expect rows of a fixed width.</p>
<h2>2. Images and Photographs</h2>
<p>Perhaps you are more used to working with images or photographs in computer vision applications.</p>
<p>Each image that you work with is itself a table structure with a width and height and one pixel value in each cell for black and white images or 3 pixel values in each cell for a color image.</p>
<p>A photo is yet another example of a matrix from linear algebra.</p>
<p>Operations on the image, such as cropping, scaling, shearing, and so on are all described using the notation and operations of linear algebra.</p>
<h2>3. One Hot Encoding</h2>
<p>Sometimes you work with categorical data in machine learning.</p>
<p>Perhaps the class labels for classification problems, or perhaps categorical input variables.</p>
<p>It is common to encode categorical variables to make them easier to work with and learn by some techniques. A popular encoding for categorical variables is the one hot encoding.</p>
<p>A one hot encoding is where a table is created to represent the variable with one column for each category and a row for each example in the dataset. A check, or one-value, is added in the column for the categorical value for a given row, and a zero-value is added to all other columns.</p>
<p>For example, the color variable with the 3 rows:</p>

		

<p>Might be encoded as:</p>

		<div id="crayon-5aa76af7156ec906828334" class="crayon-syntax crayon-theme-classic crayon-font-monaco crayon-os-pc print-yes notranslate" data-settings=" minimize scroll-mouseover">
		
			
			<p class="crayon-info"></p>
			<p class="crayon-plain-wrap"></p>
			<div class="crayon-main">
				<table class="crayon-table">
					<tr class="crayon-row">
				<td class="crayon-nums " data-settings="show">
					
				</td>
						<td class="crayon-code"><div class="crayon-pre">
<p class="crayon-line" id="crayon-5aa76af7156ec906828334-1">red, green, blue</p>
<p class="crayon-line crayon-striped-line" id="crayon-5aa76af7156ec906828334-2">1, 0, 0</p>
<p class="crayon-line" id="crayon-5aa76af7156ec906828334-3">0, 1, 0</p>
<p class="crayon-line crayon-striped-line" id="crayon-5aa76af7156ec906828334-4">0, 0, 1</p>
<p class="crayon-line" id="crayon-5aa76af7156ec906828334-5">...</p>
</div></td>
					</tr>
				</table>
			</div>
		</div>

<p>Each row is encoded as a binary vector, a vector with zero or one values and this is an example of a sparse representation, a whole sub-field of linear algebra.</p>
<h2>4. Linear Regression</h2>
<p>Linear regression is an old method from statistics for describing the relationships between variables.</p>
<p>It is often used in machine learning for predicting numerical values in simpler regression problems.</p>
<p>There are many ways to describe and solve the linear regression problem, i.e. finding a set of coefficients that when multiplied by each of the input variables and added together results in the best prediction of the output variable.</p>
<p>If you have used a machine learning tool or library, the most common way of solving linear regression is via a least squares optimization that is solved using matrix factorization methods from linear regression, such as an LU decomposition or a singular-value decomposition, or SVD.</p>
<p>Even the common way of summarizing the linear regression equation uses linear algebra notation:</p>

		

<p>Where y is the output variable A is the dataset and b are the model coefficients.</p>
<h2>5. Regularization</h2>
<p>In applied machine learning, we often seek the simplest possible models that achieve the best skill on our problem.</p>
<p>Simpler models are often better at generalizing from specific examples to unseen data.</p>
<p>In many methods that involve coefficients, such as regression methods and artificial neural networks, simpler models are often characterized by models that have smaller coefficient values.</p>
<p>A technique that is often used to encourage a model to minimize the size of coefficients while it is being fit on data is called regularization. Common implementations include the L2 and L1 forms of regularization.</p>
<p>Both of these forms of regularization are in fact a measure of the magnitude or length of the coefficients as a vector and are methods lifted directly from linear algebra called the vector norm.</p>
<h2>6. Principal Component Analysis</h2>
<p>Often, a dataset has many columns, perhaps tens, hundreds, thousands, or more.</p>
<p>Modeling data with many features is challenging, and models built from data that include irrelevant features are often less skillful than models trained from the most relevant data.</p>
<p>It is hard to know which features of the data are relevant and which are not.</p>
<p>Methods for automatically reducing the number of columns of a dataset are called dimensionality reduction, and perhaps the most popular method is called the principal component analysis, or PCA for short.</p>
<p>This method is used in machine learning to create projections of high-dimensional data for both visualization and for training models.</p>
<p>The core of the PCA method is a matrix factorization method from linear algebra. The eigendecomposition can be used and more robust implementations may use the singular-value decomposition, or SVD.</p>
<h2>7. Singular-Value Decomposition</h2>
<p>Another popular dimensionality reduction method is the singular-value decomposition method, or SVD for short.</p>
<p>As mentioned, and as the name of the method suggests, it is a matrix factorization method from the field of linear algebra.</p>
<p>It has wide use in linear algebra and can be used directly in applications such as feature selection, visualization, noise reduction, and more.</p>
<p>We will see two more cases below of using the SVD in machine learning.</p>
<h2>8. Latent Semantic Analysis</h2>
<p>In the sub-field of machine learning for working with text data called natural language processing, it is common to represent documents as large matrices of word occurrences.</p>
<p>For example, the columns of the matrix may be the known words in the vocabulary and rows may be sentences, paragraphs, pages, or documents of text with cells in the matrix marked as the count or frequency of the number of times the word occurred.</p>
<p>This is a sparse matrix representation of the text. Matrix factorization methods, such as the singular-value decomposition can be applied to this sparse matrix, which has the effect of distilling the representation down to its most relevant essence. Documents processed in this way are much easier to compare, query, and use as the basis for a supervised machine learning model.</p>
<p>This form of data preparation is called Latent Semantic Analysis, or LSA for short, and is also known by the name Latent Semantic Indexing, or LSI.</p>
<h2>9. Recommender Systems</h2>
<p>Predictive modeling problems that involve the recommendation of products are called recommender systems, a sub-field of machine learning.</p>
<p>Examples include the recommendation of books based on previous purchases and purchases by customers like you on Amazon, and the recommendation of movies and TV shows to watch based on your viewing history and viewing history of subscribers like you on Netflix.</p>
<p>The development of recommender systems is primarily concerned with linear algebra methods. A simple example is in the calculation of the similarity between sparse customer behavior vectors using distance measures such as Euclidean distance or dot products.</p>
<p>Matrix factorization methods like the singular-value decomposition are used widely in recommender systems to distill item and user data to their essence for querying and searching and comparison.</p>
<h2>10. Deep Learning</h2>
<p>Artificial neural networks are nonlinear machine learning algorithms that are inspired by elements of the information processing in the brain and have proven effective at a range of problems, not the least of which is predictive modeling.</p>
<p>Deep learning is the recent resurgence in the use of artificial neural networks with newer methods and faster hardware that allow for the development and training of larger and deeper (more layers) networks on very large datasets. Deep learning methods are routinely achieving state-of-the-art results on a range of challenging problems such as machine translation, photo captioning, speech recognition, and much more.</p>
<p>At their core, the execution of neural networks involves linear algebra data structures multiplied and added together. Scaled up to multiple dimensions, deep learning methods work with vectors, matrices, and even tensors of inputs and coefficients, where a tensor is a matrix with more than two dimensions.</p>
<p>Linear algebra is central to the description of deep learning methods via matrix notation to the implementation of deep learning methods such as Google’s TensorFlow Python library that has the word “tensor” in its name.</p>
<h2>Summary</h2>
<p>In this post, you discovered 10 common examples of machine learning that you may be familiar with that use and require linear algebra.</p>
<p>Specifically, you learned:</p>
<ul>
<li>The use of linear algebra structures when working with data such as tabular datasets and images.</li>
<li>Linear algebra concepts when working with data preparation such as one hot encoding and dimensionality reduction.</li>
<li>The ingrained use of linear algebra notation and methods in sub-fields such as deep learning, natural language processing, and recommender systems.</li>
</ul>
<p>Do you have any questions?<br>
Ask your questions in the comments below and I will do my best to answer.</p>
	</section>
	</div><br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via Machine Learning Mastery https://ift.tt/2if3vo8<br>
<br>
March 9, 2018 at 02:24AM
